File: leetcode/Ruby/No4.median_of_two_sorted_arrays.rb
=begin
Difficulty: Hard

Desc:
There are two sorted arrays nums1 and nums2 of size m and n respectively.
Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
You may assume nums1 and nums2 cannot be both empty.

Example:
nums1 = [1, 3]
nums2 = [2]
The median is 2.0

nums1 = [1, 2]
nums2 = [3, 4]
The median is (2 + 3)/2 = 2.5
=end

# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Float}
def find_median_sorted_arrays(nums1, nums2)
  i1 = 0
  i2 = 0
  list = []
  while i1 < nums1.size && i2 < nums2.size do
    num1 = nums1[i1]
    num2 = nums2[i2]
    if num1 < num2 then
      list.push(num1)
      i1 += 1
    else
      list.push(num2)
      i2 += 1
    end
  end
  if i1 < nums1.size then
    list[list.size, 0] = nums1[i1...nums1.size]
  elsif i2 < nums2.size then
    list[list.size, 0] = nums2[i2...nums2.size]
  end
  return 0.0 unless list.size > 0
  mid = list.size / 2
  return list.size % 2 === 0 ? (list[mid] + list[mid - 1]) / 2.0 : list[mid].to_f
end

# Test case
puts find_median_sorted_arrays([], [2,3])
puts find_median_sorted_arrays([], [])

File: leetcode/C#/No4.median-of-two-sorted-arrays.cs
/*
 * Difficulty:
 * Hard
 *
 * Desc:
 * There are two sorted arrays nums1 and nums2 of size m and n respectively.
 * Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
 * You may assume nums1 and nums2 cannot be both empty.
 *
 * Example 1:
 * nums1 = [1, 3]
 * nums2 = [2]
 * The median is 2.0
 *
 * Example 2:
 * nums1 = [1, 2]
 * nums2 = [3, 4]
 * The median is (2 + 3)/2 = 2.5
*/

public class Solution {
    public double FindMedianSortedArrays(int[] nums1, int[] nums2) {
        int total = nums1.Length + nums2.Length;
        int i = 0;
        int j = 0;
        int pre = 0;

        while (i + j < total / 2) {
            if (i < nums1.Length && j < nums2.Length) {
                if (nums1[i] <= nums2[j]) {
                    pre = nums1[i++];
                } else {
                    pre = nums2[j++];
                }
            } else if (i < nums1.Length) {
                pre = nums1[i++];
            } else {
                pre = nums2[j++];
            }
        }

        int next = 0;
    
        if (i < nums1.Length && j < nums2.Length) {
            next = nums1[i] <= nums2[j] ? nums1[i] : nums2[j];
        } else if (i < nums1.Length) {
            next = nums1[i];
        } else {
            next = nums2[j];
        }
        
        if (total % 2 == 1) return (double)next;
        return (pre + next) / 2f;
    }
}